sdk/core/core-websockets/src/reliableConnectionClient.ts

```diff
@@ -0,0 +1,435 @@
+   */
+  onerror: (fn: (error: unknown) => void) => void;
+}
+
+/**
+ * Options to create a reliable connection client factory.
+ */
+export interface CreateReliableConnectionOptions {
+  /**
+   * A function to check if an error is retryable.
+   */
+  isRetryable?: (err: unknown) => boolean;
+  /**
+   * Whether to swallow errors or not.
+   */
+  resolveOnUnsuccessful?: boolean;
+}
+
+/**
+ * Options to create a reliable connection client.
+ */
+export interface ReliableConnectionOptions {
+  /**
+   * The options to retry an operation.
+   */
+  retryOptions?: RetryOptions;
+  /**
+   * The identifier of the connection.
+   */
+  identifier?: string;
+}
+
+/**
+ * Creates a reliable connection client factory.
+ */
+export function createReliableConnectionClient<SendDataT, ReceiveDataT>(
+  client: ConnectionManager<SendDataT, ReceiveDataT>,
+  createOptions: CreateReliableConnectionOptions = {},
+): (opts?: ReliableConnectionOptions) => ReliableConnectionClient<SendDataT, ReceiveDataT> {
+  const { isRetryable, resolveOnUnsuccessful } = createOptions || {};
+  type WritableClient = Writable<ReliableConnectionClient<SendDataT, ReceiveDataT>>;
+  return ({
+    retryOptions: inputRetryOptions,
+    identifier,
+  }: ReliableConnectionOptions = {}): ReliableConnectionClient<SendDataT, ReceiveDataT> => {
+    const connectionId = identifier || getRandomName();
+    const retryOptions = createFullRetryOptions(inputRetryOptions);
+    let canReconnect: boolean = true;
+    function isOpenRetryable(err: unknown): boolean {
+      const res = canReconnect && (isRetryable?.(err) ?? true);
+      canReconnect = true;
+      return res;
+    }
+    let openAbortSignal: AbortSignal | undefined = undefined;
+    let openOrClosePromiseResolve: (() => void) | undefined = undefined;
+    let openOrClosePromiseReject: ((reason?: any) => void) | undefined = undefined;
+    function clearResolvers(): void {
+      openOrClosePromiseResolve = undefined;
+      openOrClosePromiseReject = undefined;
+    }
+    function rejectAndReset(reason?: any): void {
+      openOrClosePromiseReject?.(reason);
+      clearResolvers();
+    }
+    function resolveAndReset(): void {
+      openOrClosePromiseResolve?.();
+      clearResolvers();
+    }
+    const reliableConnectionClient: WritableClient = {
+      status: "disconnected",
+      open: async (openOpts: OpenOptions = {}) => {
+        switch (reliableConnectionClient.status) {
+          case "connecting":
+          case "connected": {
+            logger.info(`[${connectionId}] Client is already ${reliableConnectionClient.status}`);
+            return;
+          }
+          case "disconnecting": {
+            const errMsg = `[${connectionId}] Client is already ${reliableConnectionClient.status}`;
+            if (resolveOnUnsuccessful) {
+              logger.warning(errMsg);
+              return;
+            } else {
+              throw createError(errMsg);
```

Comment:
is it worth logging an info/verbose here as well?

Reply:
I had a lot of pain with logging in our messaging libraries so I made it a goal here to keep logging as efficient and compact as possible. In this particular case, an error is thrown so it should either show up in the final stderr stream or be caught by the application and then they have to handle/log it somehow.